pdns/dnsdist.cc

```diff
@@ -1602,6 +1748,151 @@ static void processUDPQuery(ClientState& cs, LocalHolders& holders, const struct
   }
 }
 
+#ifdef HAVE_XSK
+static void ProcessXskQuery(ClientState& cs, LocalHolders& holders, XskPacket& packet)
+{
+  uint16_t queryId = 0;
+  const auto& remote = packet.getFromAddr();
+  const auto& dest = packet.getToAddr();
+  ComboAddress proxiedRemote = remote;
+  ComboAddress proxiedDestination = dest;
+
+  try {
+    bool expectProxyProtocol = false;
+    if (!isXskQueryAcceptable(packet, cs, holders, expectProxyProtocol)) {
+      return;
+    }
+    auto query = packet.clonePacketBuffer();
```

Comment:
I think we could avoid a fair amount of the code duplication below that line, but that would require a bit of refactoring. I can do that later.